README.md
![logo](https://raw.githubusercontent.com/wiki/abntex/limarka/imagens/limarka-logo.png)

[![Build Status](https://travis-ci.org/abntex/limarka.svg?branch=master)](https://travis-ci.org/abntex/limarka) [![Coverage Status](https://coveralls.io/repos/github/abntex/limarka/badge.svg?branch=master)](https://coveralls.io/github/abntex/limarka?branch=master) [![Gem Version](https://badge.fury.io/rb/limarka.svg)](https://badge.fury.io/rb/limarka) [![Join the chat at https://gitter.im/abntex/limarka](https://badges.gitter.im/abntex/limarka.svg)](https://gitter.im/abntex/limarka?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

![Normas da ABNT em 2017: Atualizadas](https://img.shields.io/badge/Normas%20da%20ABNT%20em%202017-Atualizadas-brightgreen.svg)
![NBR 14724:2011 100%](https://img.shields.io/badge/NBR%202014724%3A2011-100%25-brightgreen.svg)

Olá! Bem vindo ao limarka!

<!-- Show what you're doing, for who, and why.  -->

O limarka é uma ferramenta que possibilita seus usuários escreverem relatórios ou trabalhos de conclusão de curso (Monografia, TCC, Dissertação ou Tese) em arquivos somente texto e **produzir PDFs em conformidades com as Normas da ABNT**. A produção desse tipo de trabalho é um momento sensível na vida dos estudantes, que pode contribuir para evasão no ensino superior. O objetivo do limarka é liberar dos estudantes os esforços de formatação e adequação às Normas da ABNT, deixando-os concentrar-se no planejamento, execução e escrita de seus trabalhos.

# Quem utiliza o limarka diz que ele é fácil e prazeroso, descubra o porquê!

<!-- Explain what makes your project special, useful, exciting! -->

* Exige conhecimentos mínimos sobre as Normas da ABNT
* Utiliza modelo com estilos pré-configurados com as normas e de fácil utilização
* Utiliza sistema para gerenciamento de citações e elaboração das referências automaticamente
* Gera o sumário automaticamente, sem esforço adicional
* Permite a escrita do trabalho em arquivos somente texto, utilizando Markdown
* As configurações são realizadas em um arquivo de formulário em PDF

## Exemplo de utilização do limarka

![logo](https://raw.githubusercontent.com/wiki/abntex/limarka/imagens/escrevendo-objetivos-em-markdown.gif)

# Como iniciar utilização

<!-- Show how to get started using or contribution to the project -->

- [Para começar a utilizar a ferramenta siga as instruções no wiki do projeto](https://github.com/abntex/limarka/wiki)
- Lembre-se de acessar o [chat no gitter](http://gitter.im/abntex/limarka) para tirar suas dúvidas sobre a ferramenta.

# Contribuindo com o projeto

<!-- State what resources and contributions you're looking for -->

Veja como é fácil contribuir!

[![OpenCollective](https://opencollective.com/limarka/backers/badge.svg)![OpenCollective](https://opencollective.com/limarka/sponsors/badge.svg)](https://opencollective.com/limarka/donate)

## Professor

- Atualize sua página pessoal/institucional adicionando uma [matéria ou link sobre o limarka](https://github.com/abntex/limarka/wiki/Imprensa)
- Solicite incluir o limarka como uma alternativa nos materiais de instruções de produção de trabalhos de conclusão, ou nas disciplinas de Metodologia.
- Ao solicitar relatórios de atividades aos alunos, sugira a utilização do limarka.
- Divulgue um notícia sobre o limarka nas listas de professores ou alunos.
- Recomende a veiculação de uma matéria sobre o limarka na agência de noticias de sua instituição.
- Sugira seus orientandos experimentarem o limarka.
- Clique na estrela no repositório do projeto (equivale a um *like*)

## Estudante

- Utilize o limarka para elaboração de relatórios
- Participe do [experimente de uso da ferramenta](https://github.com/abntex/limarka/wiki/Experimentos) - NECESSITA-SE URGENTEMENTE!
- Compartilhe alguma notícia do limarka em suas redes sociais ou lista de alunos
- Escreva seu trabalho de conclusão com o limarka e compartilhe sua experiência
- Clique na estrela no repositório do projeto (equivale a um *like*)

## Usuário Latex

- Auxilie na criação de templates latex customizados para as instituições

## Desenvolvedor Ruby

[![GitHub issues](https://img.shields.io/github/issues/abntex/limarka.svg)](https://github.com/abntex/limarka/issues)

- [Escolha uma das funcionalidades para implementar](https://github.com/abntex/limarka/issues)
- Notifique ou corrija um Bug

Acesse o [chat no gitter](http://gitter.im/abntex/limarka) e fale conosco!

## Desenvolvedor Web

Nós desejamos melhorar a experiência de utilização da ferramenta implementando
um site que gere conteúdos dinâmicos, gostaria de contribuir?

Acesse o [chat no gitter](http://gitter.im/abntex/limarka) e fale conosco!

## Todas as formas de contribuição

Consulte o arquivo [CONTRIBUTING.md](CONTRIBUTING.md) para verificar todas as formas de contribuições.

<!-- Point to other key resources, such as a contributing.md file and a roadmap. -->

## Deseja entrar em contato?

O meio de comunicação oficial do limarka é uma sala de bate papo no gitter: [![Join the chat at https://gitter.im/abntex/limarka](https://badges.gitter.im/abntex/limarka.svg)](https://gitter.im/abntex/limarka)

# Links

- A documentação do limarka será mantida no [wiki do projeto](https://github.com/abntex/limarka/wiki).
- [Vídeos](https://www.youtube.com/playlist?list=PLTnAY6TvPRKK6OgGYy3UA0oFdfCfRZesY) -- recomendo assistir em velocidade rápida
- [Convite para realização de experimento de utilização](https://github.com/abntex/limarka/wiki/Experimentos)

# Funcionalidades do limarka

- Utilização da linguagem Markdown para escrita do texto;
- Configurações através de formulário PDF ou arquivo YAML;
- Possibilidade de inclusão de códigos Latex;
- Facilidade para geração de códigos para inclusão de Figuras, Quadros, Tabelas, Referências e Citações;
- Facilidade em edição ou personalização de templates;
- Facilidade de obter suporte ao código Latex gerado (compatível com o estilo do abnTeX2);

# License

Esse gem é disponibilizado como software livre através da [Licença MIT](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
Todas as mudanças relevantes deste projeto serão documentadas neste arquivo.
O formato é baseado em [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
## 20.7.13 - 2020-07-13
### Mudanças
- Utilização de `biblatex-abnt` em vez de `abntex2-cite`  por padrão
- As configurações serão possíveis apenas a partir de arquivo YAML
- Agora o nome do pdf gerado independe do tipo do trabalho. [#196](https://github.com/abntex/limarka/issues/196)

### Melhorias implementadas
- Nova funcionalidade implementada
- Inclusão de tarefa rake de desenvolvimento `rake codelog:new[funcionalidade]`
- Configurando mirror de NY (mais rápido) para instalar o TinyTex

### Deprecated
- Funcionalidades que serão removidas em breve
- Opção `-y` será removida na próxima versão

### Removido
- Arquivo de configuração em pdf [#169](https://github.com/abntex/limarka/issues/169)
- Dependências de execução pdftk [#169](https://github.com/abntex/limarka/issues/169)
- Dependências do clipboard [#169](https://github.com/abntex/limarka/issues/169)

### Correções
- Alterações em código com erro

### Segurança
- Mudanças que corrigem vulnerabilidade

### Deploy notes
- Mudanças que impactam no processo de implantação
- Problema com uso da opção `--filtro`

---
## 19.7.3 - 2019-07-05
### Removido
- Remoção da opção de upgrade do pdf (que nunca foi implementada)

### Mudanças
- Arquivos `latexcustomizacao.sty` e referencias passaram a ser opcionais - [#194](https://github.com/abntex/limarka/issues/194)

### Melhorias implementadas
- Agora é possível ativar [outras extensões do pandoc](https://pandoc.org/MANUAL.html#non-pandoc-extensions) - [#63](https://github.com/abntex/limarka/issues/63)

---
## 19.7.2 - 2019-07-03
### Deprecated
- Comando 'check' será removido em breve

### Correções
- `help exec` passou a mostrar os parâmetros dos filtros.

---
## 19.7 - 2019-07-02
### Melhorias implementadas
- Possível utilizar filtros pandoc através das opções --filtros ou --filtros-lua [\#158](https://github.com/abntex/limarka/issues/158)

---
## 19.06 - 2019-06-22
### Melhorias implementadas
- Otimização da execução dos testes no Travis através do cache com o TinyTex entre execuções.
- Testes e deploys realizados em 2 stages no Travis (Otimização do tempo do teste e deploy)
- Os testes automatizados geram [relatório de cobertura dos testes](https://coveralls.io/github/abntex/limarka) - [\#152](https://github.com/abntex/limarka/issues/152)

### Mudanças
- Versionamento do limarka foi simplificado, agora utiliza ANO.MÊS (yy.d) para a versões oficiais, E ANO.MÊS-pre.NUMBER (yy.d-pre-n) para versões em desenvolvimento. Essa mudança permite o desenvolvimento mais ágil.
- Novas imagens do docker passaram a ser geradas no processo de build. Imagens de desenvolvimento: `limarka/ruby-latex-pandoc:dev` e `limarka/limarka:dev` Imagens de release: `limarka/limarka`, `limarka/limarka:ANO` e `limarka/limarka:ANO.MES`
- Gestão do changelog passou ser feito com o codelog - [\#191](https://github.com/abntex/limarka/issues/191)

---
<!-- Old changelog starts here -->

## [v1.1.0](https://github.com/abntex/limarka/tree/v1.1.0) (2019-06-04)
[Full Changelog](https://github.com/abntex/limarka/compare/v1.0.2...v1.1.0)

**Melhorias implementadas:**

- Atualização de testes e dependencias [\#187](https://github.com/abntex/limarka/issues/187)
- Criar capas personalizadas a partir de arquivos pdf [\#185](https://github.com/abntex/limarka/issues/185)
- Gerar listas de quadros independente através de configuração [\#181](https://github.com/abntex/limarka/issues/181)
- Deploy de imagens docker diferenciadas para TAG e desenvolvimento [\#176](https://github.com/abntex/limarka/issues/176)
- Distriuir o limarka através do docker com tinytex [\#172](https://github.com/abntex/limarka/issues/172)

## [v1.0.2](https://github.com/abntex/limarka/tree/v1.0.2) (2019-01-17)
[Full Changelog](https://github.com/abntex/limarka/compare/v1.0.1...v1.0.2)

**Melhorias implementadas:**

- Utilizar o tinytex nos testes automatizados [\#164](https://github.com/abntex/limarka/issues/164)

**Issues fechados:**

- Atualizar documentação sobre o pacote pdftk no Ubuntu 18.04 [\#168](https://github.com/abntex/limarka/issues/168)
- Atualizar documentação sobre dependências de compilação no Ubuntu 17.10.1 [\#160](https://github.com/abntex/limarka/issues/160)
- Sumário da versão v1.0.2 [\#171](https://github.com/abntex/limarka/issues/171)
- Atualizar documentação sobre o TinyTex [\#170](https://github.com/abntex/limarka/issues/170)

## [v1.0.1](https://github.com/abntex/limarka/tree/v1.0.1) (2018-10-22)
[Full Changelog](https://github.com/abntex/limarka/compare/v1.0.0...v1.0.1)

**Bugs corrigidos:**

- Limarka ainda informava que a versão 2.X era incompatível [\#167](https://github.com/abntex/limarka/issues/167)

## [v1.0.0](https://github.com/abntex/limarka/tree/v1.0.0) (2018-10-22)
[Full Changelog](https://github.com/abntex/limarka/compare/v0.8.0...v1.0.0)

**Melhorias implementadas:**

- Tornar o limarka compatível com o pandoc 2.0 [\#159](https://github.com/abntex/limarka/issues/159)

## [v0.8.0](https://github.com/abntex/limarka/tree/v0.8.0) (2018-10-20)
[Full Changelog](https://github.com/abntex/limarka/compare/v0.7.2...v0.8.0)

**Melhorias implementadas:**

- Possibilita checar e notificar problemas de dependências no sistema [\#165](https://github.com/abntex/limarka/issues/165)
- limarka --version inclui versões das dependências [\#149](https://github.com/abntex/limarka/issues/149)
- Limarka dependende de guard-shell possibilitando geração automática [\#141](https://github.com/abntex/limarka/issues/141)
- Template extraído para submódulo modelo-oficial [\#79](https://github.com/abntex/limarka/issues/79)

**Issues fechados:**

- Distribuição do limarka através de imagem Docker [\#155](https://github.com/abntex/limarka/issues/155)

## [v0.7.2](https://github.com/abntex/limarka/tree/v0.7.2) (2017-04-09)
[Full Changelog](https://github.com/abntex/limarka/compare/v0.7.1...v0.7.2)

lib/limarka/cli.rb
# coding: utf-8
require "thor"
require 'yaml'
require 'colorize'
require 'open3'
require 'limarka/configuracao'
require 'limarka/ref'
require 'limarka/conversor'
require 'terminal-table'

module Limarka

  # Essa classe é responsável por interpretar e executar os comandos de linha
  # da ferramenta. Para compreender sua utilização recomendo consultar a
  # documentação do [thor](http://whatisthor.com).
  #
  # @author Eduardo de Santana Medeiros Alexandre
  #
  class Cli < Thor
    include Thor::Actions

    default_command :exec

    @@cv = nil

    desc "check", "Verifica se o sistema está utilizando as dependências compatíveis"
    def check
      c = Limarka::Check.new()
      c.check()
    end

    # Será removida na próxima versão
    method_option :configuracao_yaml, :aliases => '-y', :type => :boolean, :desc => 'Ler configuração exportada (configuracao.yaml) em vez de configuracao.pdf', :default => true
    method_option :input_dir, :aliases => '-i', :desc => 'Diretório onde será executado a ferramenta', :default => '.'
    method_option :output_dir, :aliases => '-o', :desc => 'Diretório onde serão gerados os arquivos', :default => '.'
    method_option :compila_tex, :aliases => '-c', :desc => 'Compila arquivo tex gerando um PDF', :default => true, :type => :boolean
    method_option :templates_dir, :aliases => '-t', :desc => 'Diretório que contem a pasta templates (pandoc --data-dir)', :default => Dir.pwd
    method_option :rascunho, :aliases => '-r', :desc => 'Ler de um arquivo de rascunho em vez de "trabalho-academico.md"', :banner => "RASCUNHO_FILE"
    method_option :verbose, :aliases => '-v', :desc => 'Imprime mais detalhes da execução', :default => false, :type => :boolean
    method_option :version, :desc => 'Imprime a versão do limarka', :default => false, :type => :boolean
    method_option :filtros, :type => :array, :required => false, :desc => 'Sequencia de filtros pandoc que serão utilizados', :banner => "filtro1 filtro2 filtro3"
    method_option :filtros_lua, :type => :array, :required => false, :desc => 'Sequencia de filtros lua pandoc que serão utilizados', :banner => "f1.lua f2.lua f3.lua"

    desc "exec", "Executa o sistema para geração do documento latex e compilação"
    def exec

      if (options[:version]) then
        puts "limarka "+Limarka::VERSION
        s = `pandoc --version`
        s << `ruby --version`
        s << `latexmk --version`
        s << `xelatex --version`
        puts s
        return
      end

      #options[:output_dir] = File.absolute_path(options[:output_dir])
      Dir.chdir(options[:input_dir]) do
        t = Limarka::Trabalho.new
        t.atualiza_de_arquivos(options)
        @@cv = Limarka::Conversor.new(t,options)
        @@cv.convert
        @@cv.usa_pdftotext = false
        @@cv.compila if options[:compila_tex]
      end
    end


    desc "importa ARQUIVO", "Cria um arquivo trabalho-academico.md com o conteúdo convertido de ARQUIVO"
    long_desc "Converte documento do Word (ou similar) para trabalho-academico.md. O arquivo será criado no mesmo diretório que contém ARQUIVO. Útil quando possuímos um arquivo já digitado no word e desejamos utilizar o limarka. Mantém, por exemplo, as marcações de itálico, negrito e notas de rodapé."
    def importa(arquivo)
      diretorio = File.dirname(arquivo)
      system "pandoc", "-t", "markdown", "-o", "#{diretorio}/trabalho-academico.md", arquivo
    end

    method_option :interativo, :aliases => '-i', :desc => 'Solicita os parâmetros de forma interativa.', :type => :boolean, :default => false

#     method_option :arquivo, :aliases => '-a', :desc => 'Caminho completo ou apenas o nome do arquivo na pasta imagens. Somente arquivos existêntes são válidos. Se não for especificado e o nome do arquivo esteja copiado (na área de transferência), e o arquivo existir, ele será utilizado.'
    method_option :legenda, :aliases => '-l', :desc => 'Legenda da figura.', :default => "Legenda da figura."
    method_option :fonte, :aliases => '-f', :desc => 'Fonte da imagem.', :default => "Autor."
    method_option :rotulo, :aliases => '-r', :desc => 'Rótulo para ser utilizado na referenciação da figura, caso não especificado um será proposto.'
    method_option :dimensoes, :aliases => '-d', :desc => 'Dimensões percentuais para redimensionar a figura. Se mais de uma dimensão for especificada será apresentado um código para inclusão da imagem para cada dimensão. Útil quando deseja experimentar diversas dimensões para a Figura. Ex: 80 90 100', :default => [100], :type => :array

    long_desc <<-DESC
Esse comando imprime (1) o código para inclusão de uma figura (2) e como referenciá-la no texto. Para as figuras serem apresentadas, em conformidade com as Normas da ABNT, elas precisam serem incluídas como código Latex (abnTeX2).
DESC
    desc "fig ARQUIVO", "Imprime códigos para inclusão de imagens em conformidade com ABNT (em LaTeX)"
    def fig(arquivo=nil)

      if (arquivo) then
        if arquivo.start_with?("file://") then
          arquivo = arquivo[7,-1]
        end
      end


      if (options[:interativo]) then
        arquivo =   ask_figura_arquivo(arquivo)
        valida_figura_arquivo(arquivo) # antecipando validação issue #78
        legenda =   ask_figura_legenda
        fonte =     ask_figura_fonte
        rotulo =    ask_figura_rotulo(rotulo, arquivo)
        valida_figura_rotulo(rotulo)   # antecipando validação issue #78
        dimensoes = ask_figura_dimensoes
      else
        legenda = options[:legenda]
        fonte = options[:fonte]
        rotulo = options[:rotulo]
        if (not arquivo) then
          arquivo = ask_figura_arquivo(nil)
        end
        valida_figura_arquivo(arquivo) # antecipando validação issue #78
        rotulo = "fig:" + propoe_rotulo(File.basename arquivo, ".*") if rotulo.nil?
        valida_figura_rotulo(rotulo)   # antecipando validação issue #78
        dimensoes = options[:dimensoes]
      end




      dimensoes.each do |dim|

        escala = (dim.to_f)/100

        figura_tex = <<TEX

\\begin{figure}[htbp]
\\caption{\\label{#{rotulo}}#{legenda}}
\\begin{center}
\\includegraphics[width=#{escala}\\textwidth]{#{arquivo}}
\\end{center}
\\legend{Fonte: #{fonte}}
\\end{figure}
TEX

        say figura_tex
      end
      say "\n<!--Para referenciar a figura acima no texto utilize: Figura \\ref\{#{rotulo}} \n-->\n"
    end

    method_option :legenda, :aliases => '-l', :desc => 'Legenda da tabela.', :default => "Legenda da tabela."
    method_option :fonte, :aliases => '-f', :desc => 'Fonte da tabela.', :default => "Autor."
    method_option :rotulo, :aliases => '-r', :desc => 'Rótulo para ser utilizado na referenciação da tabela, caso não especificado um será proposto.'
    method_option :nota, :aliases => '-n', :desc => 'Texto de nota adicional. (opcional)'

    long_desc <<-DESC
Esse comando imprime duas tabela para facilitar a criação de tabelas.
DESC
    desc "tab", "Imprime códigos para inclusão de tabelas em conformidade com ABNT (em LaTeX)"
    def tab
      legenda = options[:legenda]
      fonte = options[:fonte]
      rotulo_lido = options[:rotulo]
      if rotulo_lido then
        rotulo = rotulo_lido
      else
        rotulo = "tab:"+(Time.now.to_i % 100000).to_s
      end

      nota = options[:nota]
      if nota then
        nota_linha = "  \\nota{#{nota}}%\n"
      else
        nota_linha = ""
      end

      valida_tabela_rotulo(rotulo)

      say <<TEX

\\begin{table}[htb]
\\ABNTEXfontereduzida
\\caption{#{legenda}}
\\label{#{rotulo}}
\\begin{tabular}{p{2.6cm}|p{6.0cm}|p{2.25cm}|p{3.40cm}}
  %\\hline
   \\textbf{Nível de Investigação} & \\textbf{Insumos}  & \\textbf{Sistemas de Investigação}  & \\textbf{Produtos}  \\\\
    \\hline
    Meta-nível & Filosofia\\index{filosofia} da Ciência  & Epistemologia &
    Paradigma  \\\\
    \\hline
    Nível do objeto & Paradigmas do metanível e evidências do nível inferior &
    Ciência  & Teorias e modelos \\\\
    \\hline
    Nível inferior & Modelos e métodos do nível do objeto e problemas do nível inferior & Prática & Solução de problemas  \\\\
   % \\hline
\\end{tabular}
\\legend{Fonte: #{fonte}}
\\end{table}

\\begin{table}[htb]
\\IBGEtab{%
  \\caption{#{legenda}}%
  \\label{#{rotulo}}
}{%
  \\begin{tabular}{ccc}
  \\toprule
   Nome & Nascimento & Documento \\\\
  \\midrule \\midrule
   Maria da Silva & 11/11/1111 & 111.111.111-11 \\\\
  \\midrule
   João Souza & 11/11/2111 & 211.111.111-11 \\\\
  \\midrule
   Laura Vicuña & 05/04/1891 & 3111.111.111-11 \\\\
  \\bottomrule
\\end{tabular}%
}{%
  \\fonte{#{fonte}}%
#{nota_linha}}
\\end{table}

<!--
Para referenciar a tabela acima no texto utilize: Tabela \\ref\{#{rotulo}}
-->
TEX

#        say tabela_tex


    end

    desc "cronograma", "Imprime código para facilitar elaboração de Cronograma"
    def cronograma
    # https://github.com/abntex/limarka/issues/90
    tex = <<TEX
\\begin{table}[htbp]
  \\centering
  \\caption{Cronograma de atividades}
  \\label{tab:cronograma}
  \\begin{tabular}{|c|c|c|c|c|c|}
    \\hline
    Fase & Março & Abril & Maio & Junho & Julho \\\\
    \\hline
    1 & \\textbullet & & & & \\\\
    2 & & \\textbullet & & & \\\\
    3 & & & \\textbullet & & \\\\
    4 & & & & \\textbullet & \\\\
    5 & & & & & \\textbullet \\\\
    \\hline
  \\end{tabular}
  \\legend{Fonte: Autor.}
\\end{table}

\\begin{table}[htbp]
  \\centering
  \\caption{Cronograma de atividades}
  \\label{tab:cronograma}
  \\begin{tabular}{|l|c|c|c|c|c|}
    \\hline
    Atividade & Março & Abril & Maio & Junho & Julho \\\\
    \\hline
    XXXXXXXXXXXXXXXX & \\textbullet & & & & \\\\
    XXXXXXXXXXXXXXXX & & \\textbullet & & & \\\\
    XXXXXXXX & & & \\textbullet & & \\\\
    XXXXXXXX & & & & \\textbullet & \\\\
    XXXXXXXX & & & & & \\textbullet \\\\
    \\hline
  \\end{tabular}
  \\legend{Fonte: Autor.}
\\end{table}
TEX
      puts tex
    end

    desc "menu", "Inicia um menu interativo, aceita TAB para autocompletar."
    def menu
      table = Terminal::Table.new :title => "Menu interativo - limarka v#{Limarka::VERSION}\n#{Dir.pwd}", :headings => ['Comando', 'Descrição'] do |t|
        t << ["exec", "Executa o limarka."]
        t << ["figura","Imprime código para adição de Figura."]
        t << ["tabela","Imprime código para adição de Tabela."]
        t << ["cronograma","Imprime código para inclusão de Cronograma."]
        t << ["rascunho","Executa o limarka lendo de rascunho.md"]
        t << ["web","Imprime o endereço da documentação online."]
        t << ["menu","Imprime esse menu."]
        t << ["sair","Termina o menu interativo."]
      end
      puts table
      puts "Pressione TAB para completar o comando ao digitar: 'sa'+TAB completa para 'sair'"
      sair = false
      until sair do
        cmd = ask("Qual comando deseja executar?", :limited_to => ["exec", "figura", "tabela", "cronograma","rascunho","web","menu","sair"])
        case cmd
        when "sair"
          sair = true
        when "menu"
          puts table
        when "exec"
          system "limarka","exec"
        when "web"
          puts "https://github.com/abntex/limarka/wiki"
          puts ""
        when "figura"
          system "limarka", "fig","-i"
        when "tabela"
          system "limarka", "tab"
        when "cronograma"
          system "limarka", "cronograma"
        when "rascunho"
          system "limarka", "exec", "--rascunho=rascunho.md"
        else
          "Você digitou a opção #{cmd} -- esta não é uma opção válida, apresentando o menu.\n"
          puts table
        end
      end


    end


    no_commands do

      def self.cv
        @@cv
      end

      def valida_figura_rotulo (rotulo)
        if (not rotulo =~ (/^[a-zA-Z][\w\-:]*$/)) then
          raise RuntimeError, "O rótulo não deve conter caracteres especiais. Forneça um rótulo ou remova os caracteres especiais do nome do arquivo. Rótulo atual: '#{rotulo}'"
        end
      end
      def valida_tabela_rotulo (rotulo)
        if (not rotulo =~ (/^[a-zA-Z][\w\-:]*$/)) then
          raise RuntimeError, "O rótulo não deve conter caracteres especiais, rótulo atual: #{rotulo}"
        end
      end

      def valida_figura_arquivo(arquivo)
        raise RuntimeError, "Arquivo especificado para a figura não existe: '#{arquivo}'." unless File.file?(arquivo)
      end

      def ask_figura_arquivo(arquivo = nil)
        if arquivo.nil? then
          arquivos = Dir["imagens/*"].select{ |f| File.file? f }.sort
          print_table arquivos.map.with_index{ |a, i| [i+1, *a]}
          selecao = ask("Escolha um arquivo para a Figura:").to_i
          arquivo=arquivos[selecao-1]
        end

        arquivo
      end
      def ask_figura_legenda
        legenda_padrao = "Legenda da figura."
        legenda_lida = ask("Insira o texto da legenda [#{legenda_padrao}]):")
        if legenda_lida == "" then
          legenda = legenda_padrao
        else
          legenda = legenda_lida
        end
      end

      def ask_figura_fonte
        fonte_padrao = "Autor."
        fonte_lida = ask("Insira o texto da fonte [#{fonte_padrao}]):")
        if fonte_lida == "" then
          fonte = fonte_padrao
        else
          fonte = fonte_lida
        end
      end
      def ask_figura_rotulo(rotulo, arquivo)
        # http://stackoverflow.com/questions/1268289/how-to-get-rid-of-non-ascii-characters-in-ruby

        if (not rotulo) then
          rotulo_proposto = "fig:" + propoe_rotulo(File.basename arquivo, ".*")
          rotulo_lido = ask("Rótulo para referenciar a figura [#{rotulo_proposto}]. fig:" )
          if rotulo_lido.empty? then
            rotulo = rotulo_proposto
          else
            rotulo = "fig:"+rotulo_lido
          end
        end
        rotulo
      end

      def propoe_rotulo(string_base)
        encoding_options = {
          :invalid           => :replace,  # Replace invalid byte sequences
          :undef             => :replace,  # Replace anything not defined in ASCII
          :replace           => '-',        # Use a blank for those replacements
        }

        rotulo_proposto = string_base.gsub(/\s+/, '-').encode(Encoding.find('ASCII'), encoding_options)
      end

      def ask_figura_dimensoes
        dimensoes = ask("Forneça as dimensões separadas por espaço [100]:")
        if dimensoes.empty? then
          [100]
        else
          dimensoes.split(" ")
        end
      end
    end




    desc "configuracao help", "Exporta e atualiza configurações"
    subcommand "configuracao", Limarka::Configuracao

    desc "ref", "Adiciona ou referencia bibliografia"
    subcommand "ref", Limarka::Ref

  end
end
